### PR TITLE
MAINT: combine string ufuncs by passing on auxilliary data

### PR DESCRIPTION
**Pull request by mhvk**
This ~is just a proof of concept to show~ shows how one can reduce the code by using loops for multiple functions. ~trying `maximum` and `minimum`~ EDIT: now all useful functions done.

~Right now, it is rather inelegant since I do not know how to pass on arbitrary data in the new array method framework... (@seberg?), so instead I base it on the name of the function. That said, at least that name was immediately useful!~ EDIT: it changes the array method so that it allows passing on aux data; this should really be a different PR!

cc @ngoldbaum

p.s. Shaves ~1500~ 55000 bytes off the (enormous) _multiarray_umath.cpython-311-x86_64-linux-gnu.so, though that is with debug symbols, etc.

### CODE REVIEW

**Comment by seberg** [comment]
Bytes, hehe :).  So the thing is the method object is completely opaque right now (which could hold static data).  The other options would be to expand `context` with a static data slot that can be set, I suppose.
Somehow, I didn't think static data was all that important, but folding it into `auxdata` isn't that much easier of course.  But for this type of tihng it is useful (i.e. you don't care about the singleton function call overhead at all).

**Comment by mhvk** [comment]
> the method object is completely opaque

But how would I set the method inside, say, [`init_ufunc`](https://github.com/numpy/numpy/blob/f6f3e41b9fdf3c638377aa5eaad999b4b61a2b5b/numpy/_core/src/umath/stringdtype_ufuncs.cpp#L2598)?

Also, for these purposes, it really should be extra data that goes with the specific loop, which makes `auxdata` a bit more logical. Could be part of the `spec`, I guess (which is effectively what I'm doing by using the name).

Note that in the FFT additions, I similarly made use of `auxdata` to avoid creating duplicate pieces of code...

**Comment by seberg** [comment]
So, we could add it to the `self` as a slot and create accessor (or make it public to the extend needed for that).  It is currently opaque (so even if you add it, the user can't get it).

Of course to some degree you could also prefer tagging it on to the ufunc.  But than you need to extend the way the ufuncs store loops now (since ufuncs assume evertying is in the `ArrMethod` object now).

**Comment by mhvk** [comment]
Not sure what is the best approach, but you are right that for this use case, something associated with the `ufunc` would be good enough, maybe like a "default `auxdata`" (right now, I see that in `npy_default_get_strided_loop`, the data just gets set to `NULL`). Though the loop-specific case is what would be needed for, say, `np.exp` using a generic inner loop that just calls a given function on each element (where the function likely is specific to the input dtype). I feel an ability to set `auxdata` as part of adding a given loop would be the most generally useful, i.e., it feels to me easiest would be to add a `user_data` entry to `PyArrayMethod_Spec`.

I saw that you had a note in `_dtype_api.h` that perhaps is relevant: https://github.com/numpy/numpy/blob/f6f3e41b9fdf3c638377aa5eaad999b4b61a2b5b/numpy/_core/include/numpy/_dtype_api.h#L246-L250

This refers to filling it if not `NULL`, but how in the present scheme can `auxdata`/`transferdata`/`loop_data` be anything but `NULL`?

**Comment by seberg** [comment]
You can set `auxdata` just fine, it is done in the `get_loop` function.  And `auxdata` is in fact what is used internally to pass the static `void *` into legacy ufuncs.

I.e. you can do:
```
struct {
   NpyAuxdata base;
   void *func;
} static_auxdata_struct;

/* Do nothing free, it's static */
static void
no_free(NpyAuxData *aux) {}

static_auxdata_struct min_auxdata = {
    .base={
        .free = no_free;
        // Don't bother about clone for now
    }
    func = &string_min;
}

get_loop(...) {
   if (ufunc == ...) {
       out_auxdata = min_auxdata;
   }
}
```
Which would mean you move the ufunc inspection into `get_loop`, but that's about it.

**Comment by mhvk** [comment]
That's possible, I guess, but feels wrong; `get_loop` really should not have to worry about what ufunc it is attached to; at most, it should move some bits of data from its `context` to `auxdata` (I guess that is also possible, by defining one's own `PyArrayMethod` just like is done for the legacy ufuncs...).

Note that making this a bit more general may be useful: for `strings`, the duplication of inner loops which essentially do exactly the same thing just with a different function was my main complaint (see two-but-last to-do item in gh-25693). It would really help to make this something declarative (as it was for the the legacy ufuncs). Since I think this is one of the first in-depth uses of the new array method mechanism, maybe it is still a good time to address?

**Comment by seberg** [comment]
I am very happy to add a slot to the `ArrayMethod` creation that allows adding static data.  And probably exposing it through `self`.  Internally, you could just do that.  Making it public (i.e. exposing *some* of the ArrayMethod struct), is also totally fine for obvious stuff (like this one).

Attaching things to the ufunc is plausible, but right now the ufunc has a map of `dtypes -> arraymethod`, and making that `dtypes -> (arraymethod, void *)` is a bit meeh in my eyes.

In other words, I am *very* happy if we add it to the `ArrayMethod` and also if we make it public.  We could pass it as part of `context` rather than `context->meth->static_data`, but attaching it the ufunc (which at least "feels" leaner) seems tedious.

**Comment by mhvk** [comment]
Yes, agreed, attached to the `ArrayMethod` makes most sense.

Looking at https://docs.python.org/3/c-api/type.html#c.PyType_Slot, typically a `slot` contains a function, so do you mean adding a slot for a function that provides the data? Or a slot that provides the data directly? (I think you imply the function, but just in case). And would you then use that slot (if defined), in `npy_default_get_strided_loop` (which is where currently, `out_transferdata` is set to `NULL`)?

**Comment by seberg** [comment]
> Or a slot that provides the data directly?

In almost all cases, I like functions, but I was meaning a `void *`.  A function might make sense to me if and only if:
* You use it to fill a new slot in `context`, and
* the function gets *some* input.

And I think at least the second just won't be the case (or isn't worthwhile).

**Comment by mhvk** [comment]
OK, so closer to the existing `auxdata` then!

File: numpy/_core/src/multiarray/array_method.c [hunk 1]
```diff
@@ -143,7 +149,19 @@ npy_default_get_strided_loop(
     PyArray_Descr **descrs = context->descriptors;
     PyArrayMethodObject *meth = context->method;
     *flags = meth->flags & NPY_METH_RUNTIME_FLAGS;
-    *out_transferdata = NULL;
+    if (meth->aux_data == NULL) {
+        *out_transferdata = NULL;
+    }
+    else {
+        static_auxdata_struct *transferdata = (static_auxdata_struct *)malloc(sizeof(static_auxdata_struct));
```
Comment (mhvk) [inline numpy/_core/src/multiarray/array_method.c:156]: @seberg - I followed your example above, but defining a `.clone` might actually be nicer. But, really, I'm just messing around...

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 2]
```diff
@@ -482,20 +427,24 @@ minimum_strided_loop(PyArrayMethod_Context *context, char *const data[],
         const npy_packed_static_string *sin1 = (npy_packed_static_string *)in1;
         const npy_packed_static_string *sin2 = (npy_packed_static_string *)in2;
         npy_packed_static_string *sout = (npy_packed_static_string *)out;
-        if (_compare(in1, in2, in1_descr, in2_descr) < 0) {
+        int cmp = _compare(in1, in2, in1_descr, in2_descr);
+        if (cmp == 0 && (in1 == out || in2 == out)) {
```
Comment (mhvk) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:431]: @ngoldbaum - this I think is actually a small bug in the current implementation: if the strings compare equal, then no copy should happen if either side is in-place.

**Comment by mhvk** [comment]
@seberg - I used this PR for a proof of concept of adding aux data - does it make any sense? Obviously, if it does, we should take it out of this PR...

File: numpy/_core/src/multiarray/array_method.c (hunk 1 above)
Comment (mhvk) [reply numpy/_core/src/multiarray/array_method.c:156]: @seberg - I now realize there is really no point creating this `transferdata` here since in the loop I can just access `context->method->aux_data`.  Is that what you had in mind? If so, I'll remove these changes in `array_method.*`...

**Comment by mhvk** [comment]
I went ahead and did all functions that could be fairly easily factored together... About 1000 fewer lines of code.

**Comment by mhvk** [comment]
p.s. The overall diff of `stringdtype_ufuncs.cpp` is rather dreadful. May be best to view individual commits.

**Comment by ngoldbaum** [comment]
Awesome, thanks for taking this on! I will look this over next week.

**Comment by ngoldbaum** [comment]
@lysnikolaou you probably want to look at this too

File: numpy/_core/src/multiarray/array_method.c (hunk 1 above)
Comment (seberg) [reply numpy/_core/src/multiarray/array_method.c]: Right we shouldn't build an AuxData struct at all of course, and this second iteration is what what I had in mind.  Either `context->method->const_data` or `context->const_data`, and if all we do is copy it, the first seems just as well.

(Moving it to the `context` could be done to make it a true scratch space, but we can also just add scratch space to `context` for that.)

I think we should rename `aux_data`, though.  Since there already is the other auxdata and it has a per ufunc run lifetime!  I can see `static_data`, `const_data` or even using `context`/`ctx` again.

One thing I was mentioning (but let's defer that) is that I don't want the Method objects struct fully public, so on the public side it would make sense to move this high up, so we can decide to make parts public (I think it might be fully opaque right now).

File: numpy/_core/src/multiarray/array_method.h [hunk 3]
```diff
@@ -55,6 +55,7 @@ typedef struct PyArrayMethodObject_tag {
     PyArrayMethod_StridedLoop *unaligned_strided_loop;
     PyArrayMethod_StridedLoop *unaligned_contiguous_loop;
     PyArrayMethod_StridedLoop *contiguous_indexed_loop;
+    void *aux_data;
```
Comment (seberg) [inline numpy/_core/src/multiarray/array_method.h:58]: Beyond the fact that it should be renamed, I think, the below is only relevant for followup or maybe if splitting out.

Just FYI (but that should be split out): I think I would be OK to move this up (maybe very high) and if we make things public, include the existing ones up to, and including, flags.

The "methods/functions" should *not* be public, in order to ensure we can deprecate and replace them (it would be nice to add an accessors for some so that a non-NumPy library could use our functions).
(We could also move down `name`, I really added it for debugging only.)

After adding this, it could probably also simplify the legacy arraymethod as it currently (ab)uses aux-data for static data, effectively.

**Comment by ngoldbaum** [comment]
By the way, I'm making the DType API public in another PR, so maybe wait on that one being merged for a final round on this PR because there will definitely be conflicts.

File: numpy/_core/src/multiarray/array_method.h (hunk 3 above)
Comment (mhvk) [reply numpy/_core/src/multiarray/array_method.h]: OK, thanks! I now call it `static_data` and moved it just beyond `flags`. It is still on the method, which is most logical if we allow filling it via `fill_array_method_from_slots`. 

I can see the logic of something on `context` itself, though. But maybe we need to think through a bit where exactly it will be useful...

**Comment by mhvk** [comment]
OK, this seems to work and I'm fairly happy with it. A general question is whether the dtype C API addition should be split off, and, if so, a specific one whether it might still be made part of gh-25754 (if not, then some guidance on what other files to adjust would be welcome...).

For the ufuncs themselves, the diff of the whole PR is basically unreviewable, but it's OK if done commit by commit. Happy to split it up in different PRs is that feels better.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 4]
```diff
@@ -1644,156 +1646,11 @@ string_lstrip_chars_strided_loop(
...
-        char *new_buf = (char *)PyMem_RawCalloc(s1.size, 1);
-        Buffer<ENCODING::UTF8> buf1((char *)s1.buf, s1.size);
-        Buffer<ENCODING::UTF8> buf2((char *)s2.buf, s2.size);
-        Buffer<ENCODING::UTF8> outbuf(new_buf, s1.size);
-        size_t new_buf_size = string_lrstrip_chars(
-                buf1, buf2, outbuf, STRIPTYPE::BOTHSTRIP);
-
-        if (NpyString_pack(oallocator, ops, new_buf, new_buf_size) < 0) {
-            npy_gil_error(PyExc_MemoryError, "Failed to pack string in strip");
+            npy_gil_error(PyExc_MemoryError, "Failed to pack string in ",
```
Comment (mhvk) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1149]: I fixed the missing `%s` here and below locally.

**Comment by mhvk** [comment]
@seberg - are you happy with the changes to the dtype API here? It would be nice to get this in, since other PRs can really benefit (see discussion in #25775).

**Comment by ngoldbaum** [comment]
Is there any chance we can merge this one after making the DType API public? I'm happy to resolve the inevitable merge conflicts for you on this PR if you'd prefer not to have to redo all of that.

**Comment by mhvk** [comment]
Ah, yes, you had asked that before - sorry!  Though note that what's added is really just a one-liner as far as the API is concerned:
```
    /* pointer to possible data needed inside the loop (e.g., a function pointer) */
    void *static_data;
```

Anyway, your offer to adjust is most welcome and I'm happy with any order - really, I am just hoping to have #25775 (and possibly some of the other conversions to ufuncs) benefit from my work here.

**Comment by ngoldbaum** [comment]
OK great, I'm hoping to have the dtype API PR done today so hopefully this won't take more than a couple days to sort out. Hope that's OK with you too @lysnikolaou!

No such thing as a major release of a nearly 20 year old open source project without gnrly merge conflicts when we get down to the line!

**Comment by mhvk** [comment]
Yes, n many ways it is a compliment to all those before us that there are so *few* major merge conflicts!

**Comment by seberg** [comment]
Right: For the record, I am happy with the addition, I am not 100% on how much of the `PyArrayMethod` struct we should make public, but up to there seems not unreasonable.

I was mostly hoping to delay these things until Nathans PR is in (and making the descriptor int64 and hiding the fields is gnarly business, even if hopefully not all that bad for downstream).

**Comment by lysnikolaou** [comment]
> OK great, I'm hoping to have the dtype API PR done today so hopefully this won't take more than a couple days to sort out. Hope that's OK with you too @lysnikolaou!

Sure! I'm ready to try this for fixed-length dtypes as well as soon as it's merged.

**Comment by ngoldbaum** [comment]
I don't really want to deal with making the `PyArrayMethodObject` partially public, so for now I'm making `NPY_METH_static_data` private and prepending the name with an underscore so this is mergeable without resolving the public API.

File: doc/source/reference/c-api/array.rst [hunk 5]
```diff
@@ -1711,6 +1711,17 @@ the functions that must be implemented for each slot.
...
+.. c:macro:: NPY_METH_static_data
+
+.. c:type:: void *static_data
+
+   Pointer to arbitrary data (known at compile time) that the method requires.
+   Typically used to pass in some data that allows the same loop to be used
+   for multiple methods or ufuncs. For instance, a bool could be used to
+   a string search function to indicate whether the search should be from
+   the from the front or the back, or a function pointer may be passed in
+   to do a specific test on the input.
```
Comment (mhvk) [inline doc/source/reference/c-api/array.rst:1723]: I added the new `static_data` - I hope this is the right place.

Note that this does *not* describe how to actually use the static data - as is, it is on the method object, which is opaque. Let me know what you think the best path forward is...

**Review by mhvk — COMMENTED** [review]
OK, I rebased, and things seem to work.

Main query is whether I put the documentation in the right place, and whether we should say anything about where `static_data` is exposed.

**Comment by mhvk** [comment]
Ah, oops, we clearly were pushing at the same time, sorry!  I like your suggestion - since I messed this up, let me just push from my branch...

**Comment by ngoldbaum** [comment]
Sorry for force-pushing to this branch earlier, I thought you wanted me to resolve the merge conflicts

**Comment by mhvk** [comment]
You had offered that very kindly, but I thought since I had seen the docs going in, I could just do it myself - sorry to work at cross purposes!

File: doc/source/reference/c-api/array.rst (hunk 5 above)
Comment (mhvk) [reply doc/source/reference/c-api/array.rst]: OK, so with keeping the method private, this is no longer necessary.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 2 above)
Comment (ngoldbaum) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp:432]: thanks for catching this!!

**Comment by mhvk** [comment]
OK, should be all OK now. Reminder that for review of the ufunc changes, it makes most sense to go commit by commit - the difference of the whole set is basically unintelligible.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 6]
```diff
@@ -961,48 +961,22 @@ string_intp_output_resolve_descriptors(
...
-            case IMPLEMENTED_UNARY_FUNCTIONS::STR_LEN:
-                *(T *)out = buf.num_codepoints();
-                break;
-        }
-    }
-};
-
-
-template <IMPLEMENTED_UNARY_FUNCTIONS f, const char* function_name>
+typedef bool (Buffer<ENCODING::UTF8>::*utf8_buffer_method)();
```
Comment (ngoldbaum) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:629]: nice, this is much simpler with the typedef, i'm glad the functor wasn't necessary in the end

**Review by ngoldbaum — APPROVED** [review]
Nice, always love to see -1000 LOC PRs. In my defense, sometimes when you're writing a big block of code like this it's easy to lose the forest for the trees. Thanks so much for adding the new slot in the API, it seems like an obvious nice way to avoid a ton of boilerplate. The commit-by-commit review was very straightforward. Also the tests passing strongly indicates this refactor didn't lead to any unintended behavior changes.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 6 above)
Comment (mhvk) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp:629]: This one took my by far the longest - mostly to realize that this method pointers are 16 bytes long. But, yes, once I had it, I was pleased with it!

**Comment by mhvk** [comment]
Great! I hope that with this example, refactoring the other PRs will be straightforward.

**Review by lysnikolaou — APPROVED** [review]
This was surprisingly easy to review commit-by-commit, which really speaks volumes for this approach! Let's merge this and I'll try to do it for fixed-length strings as well.

**Comment by seberg** [comment]
If you are all happy, we should just put this in, thanks.
